### ISSUE TITLE
modelsummary() adds mystery row when using robust SEs from lfe::felm()

### ISSUE THREAD

**Issue by andrewheiss** [vincentarelbundock/modelsummary#150]
This is a super edge case, but I can't figure out what's happening. When using `lfe::felm()` to estimate a model with fixed effects, `modelsummary()` adds an empty row to the top. This doesn't happen when using `fixest::feols()` to estimate the same model. 

Here's a reprex:

```r
library(tidyverse)
library(lfe)
library(fixest)
library(modelsummary)

set.seed(1234)
n <- 1000
example <- tibble(outcome = rnorm(n),
                  month = sample(1:12, n, replace = TRUE),
                  running = sample(1:50, n, replace = TRUE)) %>% 
  mutate(post = month > 6,
         treatment1 = running > 25,
         treatment2 = running > 40)

model1 <- felm(outcome ~ treatment1 * post | 
                 month + running,
               data = example)
#> Warning in chol.default(mat, pivot = TRUE, tol = tol): the matrix is either
#> rank-deficient or indefinite
model2 <- felm(outcome ~ treatment1 * post + treatment2 * post | 
                 month + running,
               data = example)
#> Warning in chol.default(mat, pivot = TRUE, tol = tol): the matrix is either
#> rank-deficient or indefinite

modelsummary(list(model1, model2))
```

<img width="357" alt="image" src="https://user-images.githubusercontent.com/73663/95691304-2243fb00-0bec-11eb-8c10-8002024b2541.png">

```r
# Mystery row at the top!
modelsummary(list(model1, model2),
             robust = TRUE)
```

<img width="355" alt="image" src="https://user-images.githubusercontent.com/73663/95691312-2cfe9000-0bec-11eb-8ed4-61a54af9f7a6.png">

```r
model3 <- feols(outcome ~ treatment1 * post | 
                  month + running,
                data = example)
#> Variables 'treatment1TRUE' and 'postTRUE' have been removed because of collinearity (see $collin.var).
model4 <- feols(outcome ~ treatment1 * post + treatment2 * post | 
                  month + running,
                data = example)
#> Variables 'treatment1TRUE', 'postTRUE' and 'treatment2TRUE' have been removed because of collinearity (see $collin.var).

modelsummary(list(model3, model4))
```

<img width="482" alt="image" src="https://user-images.githubusercontent.com/73663/95691319-46074100-0bec-11eb-9b0e-06bead415b4b.png">

```r
# No mystery row
modelsummary(list(model3, model4),
             robust = TRUE)
```

<img width="487" alt="image" src="https://user-images.githubusercontent.com/73663/95691327-54555d00-0bec-11eb-9f3f-849319b2cac0.png">

**Comment by vincentarelbundock** [comment]
Huh, this is very weird.  I need to go cook dinner, but I quickly pasted your code and I don't get any mystery row.

What version are you using?

Do you get empty rows when you call things like this:

```r
modelsummary:::extract_estimates(model3)
```

**Comment by andrewheiss** [comment]
Latest GitHub version of modelsummary. I'd been using broom 0.7.0, but in 0.7.1 they added a bunch of `felm()` changes, including a change in the argument to get robust SEs from `tidy.felm()`. It's this now:

```r
modelsummary(list(model1, model2),
             se.type = "robust")
```

…and the mystery row still appears. 

<img width="351" alt="image" src="https://user-images.githubusercontent.com/73663/95692292-bb761000-0bf2-11eb-92f6-34d831ceb0fb.png">

It doesn't appear if you use the code I originally included (`modelsummary(list(model1, model2), robust = TRUE)`) since that's deprecated now and `tidy.felm()` doesn't actually robustify anything

Running extract_estimates, though, shows what the mystery coefs are though:

```
modelsummary:::extract_estimates(model2, se.type = "robust")
```

    # A tibble: 7 x 3
    term                    statistic  value  
    <chr>                   <chr>      <chr>  
    1 treatment1TRUE          statistic1 (0.000)
    2 postTRUE                statistic1 (0.000)
    3 treatment2TRUE          statistic1 (0.000)
    4 treatment1TRUE:postTRUE estimate   -0.055 
    5 treatment1TRUE:postTRUE statistic1 (0.148)
    6 postTRUE:treatment2TRUE estimate   -0.141 
    7 postTRUE:treatment2TRUE statistic1 (0.182)

Those first three rows don't appear if you omit `se.type`

```r
modelsummary:::extract_estimates(model2)
```

    # A tibble: 4 x 3
    term                    statistic  value  
    <chr>                   <chr>      <chr>  
    1 treatment1TRUE:postTRUE estimate   -0.055 
    2 treatment1TRUE:postTRUE statistic1 (0.149)
    3 postTRUE:treatment2TRUE estimate   -0.141 
    4 postTRUE:treatment2TRUE statistic1 (0.181)

**Comment by andrewheiss** [comment]
Those first extra rows for standalone coefs don't appear when using `robust = TRUE` with `feols()` models:
```r
modelsummary:::extract_estimates(model4, robust = TRUE)

#> # A tibble: 4 x 3
#>   term                    statistic  value  
#>   <chr>                   <chr>      <chr>  
#> 1 treatment1TRUE:postTRUE estimate   -0.055 
#> 2 treatment1TRUE:postTRUE statistic1 (0.121)
#> 3 postTRUE:treatment2TRUE estimate   -0.141 
#> 4 postTRUE:treatment2TRUE statistic1 (0.223)
```

**Comment by vincentarelbundock** [comment]
OK, I see what the problem is. When we estimate a `lfe:felm` model with a rank-deficient design matrix, `tidy(mod)` will return `NaN` in all columns for the "bad" terms. However, when we call `tidy(mod, se.type="robust")`, it will return a column of "std.error" with zeros instead. Instead of omitting those terms, `modelsummary` sees those zeros and thinks they should be included.

My sense is that this should probably be fixed upstream in `broom`, but I'm curious to know what @grantmcdermott thinks (I believe he submitted the `lfe` tidyers).

Here's a MWE:


    library(lfe)
    library(broom)

    set.seed(1234)
    n <- 1000
    example <- data.frame(
      outcome = rnorm(n),
      month = sample(1:12, n, replace = TRUE),
      running = sample(1:50, n, replace = TRUE))
    example$post <- example$month > 6
    example$treatment1 <- example$running > 25
    example$treatment2 <- example$running > 40

    mod <- felm(outcome ~ treatment1 * post + treatment2 * post | 
                month + running,
                data = example)

    tidy(mod)

    #> # A tibble: 5 x 5
    #>   term                    estimate std.error statistic p.value
    #>   <chr>                      <dbl>     <dbl>     <dbl>   <dbl>
    #> 1 treatment1TRUE          NaN         NA       NaN     NaN    
    #> 2 postTRUE                NaN         NA       NaN     NaN    
    #> 3 treatment2TRUE          NaN         NA       NaN     NaN    
    #> 4 treatment1TRUE:postTRUE  -0.0554     0.149    -0.373   0.709
    #> 5 postTRUE:treatment2TRUE  -0.141      0.181    -0.776   0.438

    tidy(mod, se.type="robust")

    #> # A tibble: 5 x 5
    #>   term                    estimate std.error statistic p.value
    #>   <chr>                      <dbl>     <dbl>     <dbl>   <dbl>
    #> 1 treatment1TRUE          NaN          0       NaN     NaN    
    #> 2 postTRUE                NaN          0       NaN     NaN    
    #> 3 treatment2TRUE          NaN          0       NaN     NaN    
    #> 4 treatment1TRUE:postTRUE  -0.0554     0.148    -0.375   0.708
    #> 5 postTRUE:treatment2TRUE  -0.141      0.182    -0.771   0.441

**Comment by andrewheiss** [comment]
Ah, I see! And `fixest::feols()` (or maybe its tidier?) removes the bad terms automatically, so it works fine in `modelsummary()`

**Comment by vincentarelbundock** [comment]
Yep, that sounds right. It even seems to give you a nice error message:

```r
#> Variables 'treatment1TRUE' and 'postTRUE' have been removed because of collinearity (see $collin.var).
```

**Comment by vincentarelbundock** [comment]
So this actually goes all the way up to `lfe`:

```
> summary(mod, robust=TRUE)

Call:
   felm(formula = outcome ~ treatment1 * post + treatment2 * post |      month + running, data = example)

Residuals:
    Min      1Q  Median      3Q     Max
-3.2371 -0.6478 -0.0002  0.6151  2.9811

Coefficients:
                        Estimate Robust s.e t value Pr(>|t|)
treatment1TRUE                NA    0.00000      NA       NA
postTRUE                      NA    0.00000      NA       NA
treatment2TRUE                NA    0.00000      NA       NA
treatment1TRUE:postTRUE -0.05536    0.14777  -0.375    0.708
postTRUE:treatment2TRUE -0.14059    0.18224  -0.771    0.441

Residual standard error: 0.997 on 937 degrees of freedom
Multiple R-squared(full model): 0.06278   Adjusted R-squared: 0.0007696
Multiple R-squared(proj model): 0.001447   Adjusted R-squared: -0.06463
F-statistic(full model, *iid*):1.012 on 62 and 937 DF, p-value: 0.4519
F-statistic(proj model): 0.2713 on 5 and 937 DF, p-value: 0.9289


Warning message:
In chol.default(mat, pivot = TRUE, tol = tol) :
  the matrix is either rank-deficient or indefinite
```

**Comment by grantmcdermott** [comment]
Just saw this now, but looks like you guys have already tracked it down. Nice sleuthing.

Vincent fwiw, minor typo in your issue on the lfe repo: 

> model estimated by `feols`...

Should be `felm`.

**Comment by grantmcdermott** [comment]
PS. Andrew:

> since that's deprecated now and tidy.felm() doesn't actually robustify anything

This is probably overkill and is only tangential to the main issue in this thread but... 

`felm` always includes a robust VCV matrix as part of its return object alongside the vanilla (iid) VCV mat. This is what allows you to swap out robust SEs on the fly in the summary call. But `felm` also includes a third, "clustered" VCV mat if you specify clusters in the model formula. 

The tricky thing here is that the behavior of `summary.felm(..., robust = TRUE)` actually changes if clusters were specified. If they were, for example, then the `robust = TRUE` option actually (and I would argue counterintuitively) gives you clustered SEs rather than "robust" (i.e. heteroscedasticity consistent) SEs.

**TL;DR** I felt (feel?) strongly that this is ripe for confusion on the user side. And hence the reason I added the explicit new `se.type` argument to the felm tidiers.

(There's additional complexity here b/c the `confint.felm` method *does* allow you to consistently switch between "robust" and "clustered" SEs but it invokes different arguments. As I said, confusing!)

**Comment by vincentarelbundock** [comment]
closing because reported upstream
